File: go/tests/accelerometer/adxl362_test/main.go
//go:build linux && !tinygo

// ADXL362 hardware test — Linux host.
//
// Opens /dev/spidevB.D at 8 MHz, exercises every method of ADXL362Full,
// prints PASS/FAIL lines, and exits with the standard ===DONE: ... === summary.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/accelerometer"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	bus, err := strconv.Atoi(envOr("SPI_BUS", "0"))
	if err != nil {
		panic(err)
	}
	device, err := strconv.Atoi(envOr("SPI_DEVICE", "0"))
	if err != nil {
		panic(err)
	}

	passed, failed := 0, 0
	check := func(label string, ok bool) {
		if ok {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	conn, err := connection.NewSPIConnection(bus, device, 0, 8_000_000, nil, nil) // Create SPI connection, (bus=0, device=0, max_speed=8 MHz) → (*SPIConnection, error)
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	chip, err := accelerometer.NewADXL362Full(conn) // Create ADXL362 Full driver, (connection) → (*ADXL362Full, error)
	if err != nil {
		panic(err)
	}

	devad, devmst, partid, _, err := chip.DeviceID() // Read device IDs, () → (devid_ad, devid_mst, partid, revid, error)
	check("device_id_devid_ad", err == nil && devad == 0xAD)
	check("device_id_devid_mst", err == nil && devmst == 0x1D)
	check("device_id_partid", err == nil && partid == 0xF2)

	x, y, z, err := chip.Read() // Read 3-axis acceleration, () → (x, y, z g, error)
	check("read_12bit", err == nil)
	check("read_12bit_floats", err == nil && !floatIsZero(x, y, z))

	_, _, _, err = chip.Read8bit() // Read 8-bit acceleration, () → (x, y, z g, error)
	check("read_8bit", err == nil)

	_, err = chip.Temperature() // Read temperature, () → (°C, error)
	check("temperature", err == nil)

	chip.SetRange(4)                          // Set measurement range, (rangeG=4) → error
	chip.SetODR(200.0)                        // Set output data rate, (odrHz=200.0) → error
	chip.SetHalfBandwidth(true)               // Set antialiasing bandwidth, (enabled=true) → error
	chip.SetNoiseMode(accelerometer.ADXL362NoiseLow) // Set noise mode, (mode=NOISE_LOW=1) → error
	check("set_range_odr_noise", true)

	_, err = chip.Status() // Read STATUS register, () → (byte, error)
	check("status", err == nil)
	_, err = chip.Awake() // Check AWAKE bit, () → (bool, error)
	check("awake", err == nil)
	_, err = chip.DataReady() // Check DATA_READY, () → (bool, error)
	check("data_ready", err == nil)
	_, err = chip.FifoEntries() // Read FIFO entry count, () → (uint16, error)
	check("fifo_entries", err == nil)

	chip.ConfigureFifo(accelerometer.ADXL362FifoStream, false, 128)        // Configure FIFO, (mode=STREAM=2, storeTemp=false, watermark=128) → error
	chip.SetActivityThreshold(0.5, true)                                     // Set activity threshold, (thresholdG=0.5, referenced=true) → error
	chip.SetActivityTime(5)                                                   // Set activity time, (samples=5) → error
	chip.SetInactivityThreshold(0.2, true)                                    // Set inactivity threshold, (thresholdG=0.2, referenced=true) → error
	chip.SetInactivityTime(30)                                                 // Set inactivity time, (samples=30) → error
	chip.EnableActivityDetection(true)                                        // Enable activity detection, (enabled=true) → error
	chip.EnableInactivityDetection(true)                                      // Enable inactivity detection, (enabled=true) → error
	chip.SetLinkLoopMode(accelerometer.ADXL362LinkLoopLoop)                   // Set link/loop mode, (mode=LOOP=3) → error
	check("activity_inactivity_config", true)

	chip.SetInterrupt(1, accelerometer.ADXL362SourceDataReady, true)         // Map DATA_READY to INT1, (pin=1, source=DATA_READY=0, enabled=true) → error
	chip.SetInterrupt(2, accelerometer.ADXL362SourceAwake, true)              // Map AWAKE to INT2, (pin=2, source=AWAKE=6, enabled=true) → error
	chip.SetInterruptPolarity(1, true)                                        // Set INT1 active-low, (pin=1, activeLow=true) → error
	check("interrupt_mapping", true)

	chip.SelfTest(true)                                                        // Enable self-test, (enabled=true) → error
	chip.SelfTest(false)                                                       // Disable self-test, (enabled=false) → error
	check("self_test", true)

	chip.SoftReset()                                                           // Soft-reset the chip, () → error
	check("soft_reset", true)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func floatIsZero(values ...float32) bool {
	for _, v := range values {
		if v != 0 {
			return false
		}
	}
	return false
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}

File: rust/tests/accelerometer/adxl362_test/src/main.rs
use embedded_hal_bus::spi::ExclusiveDevice;
use linux_embedded_hal::SpidevBus;
use periph::chips::accelerometer::{
    Adxl362Full, ADXL362_FIFO_STREAM, LINKLOOP_LOOP, NOISE_LOW, SOURCE_AWAKE, SOURCE_DATA_READY,
};
use spidev::{SpiModeFlags, Spidev, SpidevOptions};

struct NullCs;
impl embedded_hal::digital::ErrorType for NullCs {
    type Error = core::convert::Infallible;
}
impl embedded_hal::digital::OutputPin for NullCs {
    fn set_low(&mut self) -> Result<(), Self::Error> { Ok(()) }
    fn set_high(&mut self) -> Result<(), Self::Error> { Ok(()) }
}

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let bus: u8 = std::env::var("SPI_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(0);
    let dev: u8 = std::env::var("SPI_DEVICE").ok().and_then(|v| v.parse().ok()).unwrap_or(0);
    let dev_path = format!("/dev/spidev{}.{}", bus, dev);

    let mut passed = 0i32;
    let mut failed = 0i32;

    let mut spi = Spidev::open(dev_path).expect("open spi");
    spi.configure(&SpidevOptions::new()
        .max_speed_hz(8_000_000)
        .mode(SpiModeFlags::SPI_MODE_0)
        .build()).expect("configure spi");
    let bus_obj = SpidevBus(spi);
    let device = ExclusiveDevice::new_no_delay(bus_obj, NullCs).expect("spi device");

    let mut chip = Adxl362Full::new(device).expect("init ADXL362");    // Create ADXL362 full driver, (spi) → Result

    let ids = chip.device_id().expect("device_id");                      // Read device IDs, () → Result<(DEVID_AD, DEVID_MST, PARTID, REVID)>
    check_true!(ids.0 == 0xAD, "device_id_devid_ad", passed, failed);    // verify DEVID_AD = 0xAD
    check_true!(ids.1 == 0x1D, "device_id_devid_mst", passed, failed);  // verify DEVID_MST = 0x1D
    check_true!(ids.2 == 0xF2, "device_id_partid", passed, failed);      // verify PARTID = 0xF2

    let r12 = chip.read().expect("read");                                // Read 12-bit acceleration, () → Result<(f32, f32, f32)> g
    check_true!(true, "read_12bit", passed, failed);

    let r8 = chip.read_8bit().expect("read_8bit");                       // Read 8-bit acceleration, () → Result<(f32, f32, f32)> g
    check_true!(true, "read_8bit", passed, failed);

    let _t = chip.temperature().expect("temperature");                   // Read temperature, () → Result<f32> °C
    check_true!(true, "temperature", passed, failed);

    chip.set_range(4).expect("set_range(4)");                            // Set measurement range, (range_g=4) → Result<()>
    chip.set_odr(200.0).expect("set_odr(200)");                          // Set output data rate, (odr_hz=200.0) → Result<()>
    chip.set_half_bandwidth(true).expect("set_half_bandwidth(true)");    // Set antialiasing bandwidth, (enabled=true) → Result<()>
    chip.set_noise_mode(NOISE_LOW).expect("set_noise_mode(LOW)");        // Set noise mode, (mode=NOISE_LOW) → Result<()>
    check_true!(true, "set_range_odr_noise", passed, failed);

    let _status = chip.status().expect("status");                        // Read STATUS register, () → Result<u8>
    let _awake = chip.awake().expect("awake");                           // Check AWAKE bit, () → Result<bool>
    let _dr = chip.data_ready().expect("data_ready");                    // Check DATA_READY, () → Result<bool>
    let _n = chip.fifo_entries().expect("fifo_entries");                 // Read FIFO entry count, () → Result<u16>
    check_true!(true, "status_awake_dr_fifo", passed, failed);

    chip.configure_fifo(ADXL362_FIFO_STREAM, false, 128).expect("configure_fifo");  // Configure FIFO, (mode=STREAM, store_temp=false, watermark=128) → Result<()>
    chip.set_activity_threshold(0.5, true).expect("set_activity_threshold(0.5, true)");  // Set activity threshold, (threshold_g=0.5, referenced=true) → Result<()>
    chip.set_activity_time(5).expect("set_activity_time(5)");            // Set activity time, (samples=5) → Result<()>
    chip.set_inactivity_threshold(0.2, true).expect("set_inactivity_threshold(0.2, true)");  // Set inactivity threshold, (threshold_g=0.2, referenced=true) → Result<()>
    chip.set_inactivity_time(30).expect("set_inactivity_time(30)");      // Set inactivity time, (samples=30) → Result<()>
    chip.enable_activity_detection(true).expect("enable_activity_detection(true)");      // Enable activity detection, (enabled=true) → Result<()>
    chip.enable_inactivity_detection(true).expect("enable_inactivity_detection(true)");  // Enable inactivity detection, (enabled=true) → Result<()>
    chip.set_link_loop_mode(LINKLOOP_LOOP).expect("set_link_loop_mode(LOOP)");            // Set link/loop mode, (mode=LOOP) → Result<()>
    check_true!(true, "activity_inactivity_config", passed, failed);

    chip.set_interrupt(1, SOURCE_DATA_READY, true).expect("set_interrupt(1, DATA_READY, true)");  // Map DATA_READY to INT1, (pin=1, source=DATA_READY, enabled=true) → Result<()>
    chip.set_interrupt(2, SOURCE_AWAKE, true).expect("set_interrupt(2, AWAKE, true)");             // Map AWAKE to INT2, (pin=2, source=AWAKE, enabled=true) → Result<()>
    chip.set_interrupt_polarity(1, true).expect("set_interrupt_polarity(1, true)");              // Set INT1 active-low, (pin=1, active_low=true) → Result<()>
    check_true!(true, "interrupt_mapping", passed, failed);

    chip.self_test(true).expect("self_test(true)");                      // Enable self-test, (enabled=true) → Result<()>
    chip.self_test(false).expect("self_test(false)");                    // Disable self-test, (enabled=false) → Result<()>
    check_true!(true, "self_test", passed, failed);

    chip.soft_reset().expect("soft_reset");                              // Soft-reset the chip, () → Result<()>
    check_true!(true, "soft_reset", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}